### ISSUE TITLE
Incus from source on Rocky Linux issues with spice and qemu-kvm

### ISSUE THREAD

**Issue by lebonez** [lxc/incus#1301]
# Required information

 * Distribution: Rocky Linux
 * Distribution version: 9.4
 * The output of "incus info" or if that fails: 
   * Kernel version: 5.14.0-427.37.1.el9_4.x86_64
   * LXC version: 
   * Incus version: 6.6
   * Storage backend in use: Ceph

# Issue description

High likelihood this could be a duplicate, I've built Incus on Rocky Linux 9.4 from source. I am hitting two major issues. It seems that Incus is expecting qemu-system-x86_64 to exist on RedHat derivatives but this does not exist it exists at /usr/libexec/qemu-kvm. Doing a symlink from `ln -s /usr/libexec/qemu-kvm /usr/bin/qemu-system-x86_64` works around it but that isn't ideal. The next major issue is the drop of Spice support in RedHat 9. Is it possible to build incus without expecting spice? 

# Steps to reproduce

 1. Install Rocky Linux 9.4
 2. Build from source
 3. Attempt to run a VM like `lxc launch images:rockylinux/9/cloud --vm` VM feature won't be supported even though it should be"
 4. `ln -s /usr/libexec/qemu-kvm /usr/bin/qemu-system-x86_64`
 5. `lxc launch images:rockylinux/9/cloud --vm` now it will complain about the VM configs due to the spice settings not being part of qemu anymore in RedHat 9.

# Questions

* Is it possible in the future that incus will support the /usr/libexec/qemu-kvm location?
* Is it also possible that incus will dynamically swap away from spice as needed if it doesn't exist?

We are definitely awaiting a RedHat 9 zabbly incus LTS.

**Comment by stgraber** [comment]
1) Should be possible though that's not particularly great as we normally want to use the architecture specific QEMU binaries. How does RHEL handle the case where you have the QEMU system binaries for multiple architectures installed? Is qemu-kvm some kind of shell script or the actual QEMU binary? 

2) We may be able to detect the lack of SPICE support and degrade things, though that means no VGA access for your VMs which will make some use cases pretty unpleasant (like running Windows or a Linux desktop VM).

**Comment by lebonez** [comment]
Thank you for the quick response.

1) Indeed /usr/libexec/qemu-kvm on redhat is just the qemu binary unsure if it differentiates on architectures. Looking at the qemu packages available on EL9 there are no extra architecture packages like there used to be on EL8 and older.

2) Yes lack of Spice support seems most complicated but doesn't have a good solution like symlinking or a simple patch. Generally we still have virt-viewer available on EL9 but not spicy or remote-viewer. Which I assume means just have to drop to VNC only? I think the best workaround for my situation would be if there is a way to disabling spice support as a config flag or something. At a later time supporting virt-viewer VNC could be useful.

**Comment by stgraber** [comment]
We're not likely to add support for VNC as we actively use features of SPICE which VNC doesn't support, including USB passthrough, video acceleration channels, multi display support, ...

But hopefully we can degrade things in a way which just prevents attaching to the VGA console altogether on such systems.

**Comment by mmol67** [comment]
Hello! Maybe you can use https://copr.fedorainfracloud.org/coprs/ligenix/enterprise-qemu-spice/
This was working for me in Rocky 9.4 and LXD. Now I'm getting the same complain:
`Error: Failed instance creation: Failed creating instance record: Instance type "virtual-machine" is not supported on this server: QEMU failed to run feature checks`
But maybe someone with more knowledge than me can fix it using this repository.

Discussion in Rocky Linux forum: https://forums.rockylinux.org/t/spice-support-was-dropped-in-rhel-9/6753/41?page=2

Regards.

**Comment by damex** [comment]
@mmol67 i think thats not a good solution to a 'self made problem where spice is forced'.

@stgraber would it be acceptable to add something like spice = true/false to a instance/profile config section and wrap spice requirements against it? i can try to come up with something. 'rhel' (almalinux?) usually used on servers where spice features are not really a usual occurrence. there should some subtitudes with for some components

**Comment by stgraber** [comment]
As I said above, if we can reliably detect it then we can degrade by not having the VGA console on those plarforms. That will certainly impact the usability of VMs as you'll be limited to just the text console, making running things like Windows VMs pretty challenging. 

You should also never cluster any such systems with a system that has proper spice support as the difference in QEMU config would cause an instant VM crash during live migration.

**Comment by damex** [comment]
@stgraber i see that you want a way to detect it automatically, but why not a simple spice=true/false flag at the instance or profile config section?
no magic detection with explicit definition and vm with spice=false migrates freely across a mixed cluster without the scheduler needing to probe host QEMU capabilities.

would that be acceptable?

windows vms can get standard vga and some things along with it instead of spice

**Comment by damex** [comment]
well, if we start guessing qemu capabilities, vm config would need to change per host... that's a huge issue down the line

**Comment by damex** [comment]
https://github.com/lxc/incus/pull/3099/changes maybe this would suffice? current behavior stays and we get an option to disable spice flag along with automatically generated spice devices.

**Comment by stgraber** [comment]
As mentioned in the PR, we really don't want to make this user configurable.
Instead we should have `checkFeatures` detect it and the restrict the VMs accordingly.

**Comment by damex** [comment]
@stgraber how do you suggest it is gonna be handled on clusters with machines that has spice and that don't have support for spice?

I think flag is a good step since it provides an option (currently there are NONE) to stop guessing and provide consistent configuration, even when you have support for spice but do not want or need it due to compliance reasons.

**Comment by damex** [comment]
I currently override every single one spice device including qemu wrapper that ignores -spice option.

That ... 'works'. I hope there is anything better. I think explicit is better and already made a pull request that have it via config option but it got rejected. Anything else i could do ?

**Comment by stgraber** [comment]
> [@stgraber](https://github.com/stgraber) how do you suggest it is gonna be handled on clusters with machines that has spice and that don't have support for spice?

You should create two cluster groups, one for the machines with SPICE, one with the machines without them, same you would with CPU generations.

**Comment by damex** [comment]
How would cpu generation relate to spice support?

Let's say amd64 whatever v2 and v4. How does that help?

Or you want to differentiate using cluster groups? Like one group has spice and other don't? Then we need fixed config option, no? Could i make a cluster wide option ?

**Comment by stgraber** [comment]
It doesn't. I'm saying it's the same solution for both. You use cluster groups to group servers together based on CPU generations so VMs can be migrated properly within that set of server.

You'd create cluster groups with servers that have SPICE and servers that don't for the same reason, the VMs will be allowed to migrate within their group.

**Comment by damex** [comment]
Can you point to the right direction of implementation? Maybe what to check inside codebase or something. I will check what i could do this weekend

**Comment by stgraber** [comment]
https://github.com/lxc/incus/blob/main/internal/server/instance/drivers/driver_qemu.go#L10248 needs a check to see if Spice is supported by QEMU.

Easiest would likely be to add `query-spice` as a command to internal/server/instance/drivers/qmp/, then the `checkFeatures` code can call `query-spice` and see what happens. On a normal system it should give the response struct from https://qemu-project.gitlab.io/qemu/interop/qemu-qmp-ref.html#command-QMP-ui.query-spice or on systems lacking support, you should get an error that the command doesn't exist.

Then from the checkFeatures code, you can call that QMP command and if it fails, then mark the spice features as missing. Finally in the code that sets up spice, skip if the feature is reported as missing.

**Comment by damex** [comment]
> https://github.com/lxc/incus/blob/main/internal/server/instance/drivers/driver_qemu.go#L10248 needs a check to see if Spice is supported by QEMU.
> 
> Easiest would likely be to add `query-spice` as a command to internal/server/instance/drivers/qmp/, then the `checkFeatures` code can call `query-spice` and see what happens. On a normal system it should give the response struct from https://qemu-project.gitlab.io/qemu/interop/qemu-qmp-ref.html#command-QMP-ui.query-spice or on systems lacking support, you should get an error that the command doesn't exist.
> 
> Then from the checkFeatures code, you can call that QMP command and if it fails, then mark the spice features as missing. Finally in the code that sets up spice, skip if the feature is reported as missing.

ins't that's chicken and egg problem? it has to start with spice to query it?

https://qemu-project.gitlab.io/qemu/interop/qemu-qmp-ref.html#command-QMP-char.query-chardev-backends maybe this woudl work checking for ... spiceport or something like that ?

**Comment by stgraber** [comment]
> ins't that's chicken and egg problem? it has to start with spice to query it?

The QEMU started by checkFeatures doesn't currently pass the SPICE arguments, so it should be able to give you a QMP endpoint that you can attempt the query-spice against, then that command should fail as not available if QEMU is missing support, or will return as existing but server disabled if QEMU does have support.

**Comment by damex** [comment]
@stgraber https://github.com/lxc/incus/pull/3103 like this? it was pretty straightforward to add. is there anything else i could/have missed?

i did test it on rhel9.6 (running bluefield2 (arm64 dpu with 8 core and 16gb ram)), rhel10/alma10 x86_64 boxes and it work just fine as it is.

**Comment by damex** [comment]
can we also wrap plan9 usage like we did with spice ?

https://github.com/lxc/incus/blob/main/internal/server/instance/drivers/driver_qemu.go#L4048 here we have only windows excluded while rhel does not build plan9 support in.

@stgraber would it be acceptable to do the same with 'plan9' feature detection? there is no separate check for plan9 present but maybe we can try to look for device-list-properties of virtio-9p-pci ?

**Comment by stgraber** [comment]
Yeah, if we can detect missing 9p support in QEMU in checkFeatures, then we can similarly skip those devices.

**Comment by stgraber** [comment]
Not sure why 9p wasn't mentioned in this issue. The only two things we had here were spice and the QEMU binary path that I sent a separate PR for.

**Comment by damex** [comment]
yeah, not sure. i found the thread because of 'spice' and spice issue seem to be 'solved' so far :)

want me to test qemu binary fix?

i checked and something like this for plan9:
1. qmp detection
2. wrap config/agent block based on presence on 9p 
3. gate io.bus=9p for disk devices

i have been running those on rhel10 host and it seem to work just fine (no side effects noticed). for example mikrotik chr does not need nor support this :)

```
      raw.qemu.conf: |-
        [fsdev "qemu_config"]
        [device "dev-qemu_config-drive-9p"]
        [fsdev "qemu_agent"]
        [device "dev-qemu_agent-drive-9p"]
```


on a side note there might be a need for virtiofsd fallback (support?) here for agent drive if we want to preserve 'current behavior with agent passing via plan9' when plan9 is actually absent... but i am not sure that should be mixed together right now. and virtiofsd might actually be absent too :)

**Comment by damex** [comment]
added https://github.com/lxc/incus/pull/3110 since it 'got lost' before last changes got merged. qemu audio is SPICE specific and we get this device block otherwise
```
[device "qemu_spice-audio"]
driver = "virtio-sound-pci"
bus = "qemu_pcie0"
addr = "00.7"
audiodev = "qemu_spice-audiodev"
```

also added https://github.com/lxc/incus/pull/3109 this pull request for feature gating plan9.

tested both on multiple rhel9 and rhel10 boxes

ps could we please get .1 minor release bump if those changes along with qemu binary path get merged stating introduction rhel support?

**Comment by stgraber** [comment]
As briefly mentioned in the PR.

virtio-sound shouldn't have to be tied to spice and there may be legitimate reasons for wanting an audio target in a VM even if it can't make it back to the client, mostly to appease applications that may require an audio target. 

It would probably be worth checking if the device can be set up without the audiodev part and if so, rename it to just qemu_sound or something and only make the attachment to spice be what changes based on qemu support.

**Comment by damex** [comment]
@stgraber adjusted but... that the approach not gonna work for rhel.

```
qemu-system-x86_64:/run/incus/chr10/qemu.conf:160: 'virtio-sound-pci' is not a valid device model name
```

should i gate 'audio' or 'sound' as a feature via availability of virtio-sound-pci?

rhel10 has those:
```
Sound devices:
name "hda-duplex", bus HDA, desc "HDA Audio Codec, duplex (line-out, line-in)"
name "hda-micro", bus HDA, desc "HDA Audio Codec, duplex (speaker, microphone)"
name "hda-output", bus HDA, desc "HDA Audio Codec, output-only (line-out)"
name "ich9-intel-hda", bus PCI, desc "Intel HD Audio Controller (ich9)"
name "intel-hda", bus PCI, desc "Intel HD Audio Controller (ich6)
```

or maybe we degrade it to one of those?

**Comment by stgraber** [comment]
We don't generally like full hardware emulation as those code paths are were the vast majority of security issues with QEMU tend to be. So it would probably be best to add more feature detection logic for this and skip the device when virtio-sound-pci is missing.

I'd probably call the feature `virtio-sound`.

**Comment by damex** [comment]
https://github.com/lxc/incus/pull/3115 here it is. made 'virtio-sound' based on presence of virtio-sound-pci.

whew... hopefully that's all :)

**Comment by damex** [comment]
well, ofcourse tested on rhel9/rhel10 and it works
